controller/gateway/controller_reconciler_utils.go

```diff
@@ -875,10 +876,25 @@ func getSupportedKindsWithResolvedRefsCondition(ctx context.Context, c client.Cl
 				secretNamespace = string(*certificateRef.Namespace)
 			}
 
+			isReferenceGranted := true
+			// In case there is a cross-namespace reference, check if there is any referenceGrant allowing it.
+			if secretNamespace != gatewayNamespace {
+				referenceGrantList := &gatewayv1beta1.ReferenceGrantList{}
+				err = c.List(ctx, referenceGrantList, client.InNamespace(secretNamespace))
+				if err != nil {
+					return
```

Comment:
As named return values are not something we love relying on, I think it's better to always explicitly define what's to be returned. Wrapping the error might make sense as well here.
```suggestion
					return nil, metav1.Condition{}, fmt.Errorf("failed to list ReferenceGrants: %w", err)
```